sensor/common/clusterentities/store_ips.go

```diff
@@ -0,0 +1,244 @@
+		}
+	}
+	for deploymentID, data := range updates {
+		if data.isDeleteOnly() {
+			// A call to Apply() with empty payload of the updates map (no values) is meant to be a delete operation.
+			continue
+		}
+		e.applySingleNoLock(deploymentID, *data)
+	}
+}
+
+func (e *podIPsStore) purgeDeploymentNoLock(deploymentID string) {
+	if e.historyEnabled() {
+		e.moveDeploymentToHistory(deploymentID)
+	} else {
+		e.deleteDeploymentFromCurrent(deploymentID)
+		// In case we allow in the future to disable history during runtime, we would need to remove here all
+		// expired data for deploymentID from the history.
+	}
+}
+
+func (e *podIPsStore) LookupByNetAddr(ip net.IPAddress, port uint16) (results, historical []LookupResult) {
+	e.mutex.RLock()
+	defer e.mutex.RUnlock()
+	for deploymentID := range e.ipMap[ip] {
+		result := LookupResult{
+			Entity:         networkgraph.EntityForDeployment(deploymentID),
+			ContainerPorts: []uint16{port},
+		}
+		results = append(results, result)
+	}
+	// if there is a match in the map, then there is no need to search in history,
+	// as it may contain data about different past deployment using this address
+	for histDeploymentID := range e.historicalIPs[ip] {
+		result := LookupResult{
+			Entity:         networkgraph.EntityForDeployment(histDeploymentID),
+			ContainerPorts: []uint16{port},
+		}
+		historical = append(historical, result)
+	}
+	return results, historical
+}
+
+func (e *podIPsStore) applySingleNoLock(deploymentID string, data EntityData) {
+	ipsSet := e.reverseIPMap[deploymentID].Unfreeze()
+	for ip := range data.ips {
+		ipsSet.Add(ip)
+
+		// check if this IP already belongs to other deployment
+		deplSet := e.ipMap[ip]
+		deplSet.Add(deploymentID)
+		// This IP has more than one deployment! Interesting, let's record it.
+		if deplSet.Cardinality() > 1 {
+			metrics.ObserveManyDeploymentsSharingSingleIP(ip.AsNetIP().String(), deplSet.AsSlice())
+		}
+		e.ipMap[ip] = deplSet
+		// If the IP being currently added was already in history,
+		// we must remove it from the history to prevent expiration after a while.
+		_ = e.deleteFromHistory(deploymentID, ip)
+	}
+
+	e.reverseIPMap[deploymentID] = ipsSet.Freeze()
+}
+
+// moveDeploymentToHistory is a convenience function that removes data from the current map and adds it to history
+func (e *podIPsStore) moveDeploymentToHistory(deploymentID string) {
+	e.addToHistory(deploymentID)
+	e.deleteDeploymentFromCurrent(deploymentID)
+}
+
+func (e *podIPsStore) addToHistory(deploymentID string) {
+	ipSet := e.reverseIPMap[deploymentID]
+	for _, ip := range ipSet.AsSlice() {
+		if _, ok := e.historicalIPs[ip]; !ok {
+			e.historicalIPs[ip] = make(map[string]*entityStatus)
+		}
+		e.historicalIPs[ip][deploymentID] = newHistoricalEntity(e.memorySize)
+	}
+}
+
+// deleteDeploymentFromCurrent deletes all data for given deployment from the current map
+func (e *podIPsStore) deleteDeploymentFromCurrent(deploymentID string) set.FrozenSet[net.IPAddress] {
```

Comment:
I think the returned set is not used anywhere.

Reply:
it used to be, believe me :D Sorry for so many leftovers 😅